Math/NumberTheory/Primes/Factorisation/QuadraticSieve.hs

```diff
@@ -192,41 +193,39 @@ findSquares n (QuadraticSieveConfig t m k h) = runST $ do
           smoothLogSieveM sievingLogIntervalM factorBaseWithSquareRoots a b c m
           sievedLogInterval <- U.unsafeFreeze sievingLogIntervalM
           let
-            -- This removes duplicates. This is an issue only when trying to factor small numbers.
             newSmoothNumbers = findLogSmoothNumbers factorBase m h decompositionOfA b sievingInterval sievedLogInterval
-            smoothNumbers = previousSmoothNumbers `M.union` newSmoothNumbers
-            matrixSmoothNumbers
+            currentSmoothNumbers = previousSmoothNumbers `M.union` newSmoothNumbers
+            smoothNumbers
               -- This traces the number of smooth numbers and primes found in the previous sieving block.
-              | trace ("Smooth Numbers: " ++ show (length mat) ++ "\nPrimes: " ++ show numberOfConstraints) False = undefined
+              | trace ("Smooth Numbers: " ++ show (M.size currentSmoothNumbers) ++ "\nPrimes: " ++ show numberOfConstraints) False = undefined
               -- Enough smooth numbers are found. Not too many smooth numbers are
               -- taken to ensure the dimension of the matrix is not too large.
-              | numberOfConstraints < length mat = take (numberOfConstraints + 3 * (k + 2)) $ M.assocs smoothNumbers
+              | numberOfConstraints < M.size currentSmoothNumbers = M.take (numberOfConstraints + 5 * (k + 1)) currentSmoothNumbers
               -- More smooth numbers are needed.
-              | otherwise                        = goSelfInitSieving smoothNumbers otherCoeffs
+              | otherwise                        = goSelfInitSieving currentSmoothNumbers otherCoeffs
               where
-                numberOfConstraints = S.size $ foldMap convertToSet mat
                 -- The tracing prints how many numbers satisfy the given threshold and how many of them are actually smoooth.
-                mat = trace ("Log Filtering: " ++ show (U.length (U.filter (<= h) sievedLogInterval)) ++ " -> " ++ show (M.size newSmoothNumbers)) $
-                  M.elems smoothNumbers
-          pure matrixSmoothNumbers
+                numberOfConstraints = trace ("Log Filtering: " ++ show (U.length (U.filter (<= h) sievedLogInterval)) ++ " -> " ++ show (M.size newSmoothNumbers)) $
+                  IS.size $ foldMap setOddPowers currentSmoothNumbers
+          -- Removes columns which could never be part of a solution.
+          pure $ removeColumns 0 smoothNumbers
 
-    -- Removes columns which could never be part of a solution.
-    sievingData = removeRows $ goSieving mempty initialDecompositionOfA
-    matrix = translate $ fmap (convertToSet . snd) sievingData
+    sievingData = goSieving mempty initialDecompositionOfA
+    matrix = translate . M.elems $ M.map setOddPowers sievingData
```

Comment:
`M.elems . M.map` does not make much sense, I think. It is better to extract elements first, and apply `setOddPower` later.